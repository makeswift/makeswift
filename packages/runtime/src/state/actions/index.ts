import { isAction } from '@reduxjs/toolkit'

import { type BuilderAction, BuilderActionTypes } from '../builder-api'
import { type HostAction, HostActionTypes } from '../host-api'

import { type InternalAction, InternalActionTypes } from './internal'

export { type UnknownAction } from '@reduxjs/toolkit'
export { type DocumentPayload } from '../shared-api'

export const ActionTypes = {
  ...HostActionTypes,
  ...BuilderActionTypes,
  ...InternalActionTypes,
} as const

export type Action = HostAction | BuilderAction | InternalAction

export function isKnownAction(action: unknown): action is Action {
  return isAction(action) && Object.hasOwn(ActionTypes, action.type)
}
