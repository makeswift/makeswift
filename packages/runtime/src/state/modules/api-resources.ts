import {
  APIResourceType,
  APIResourceLocale,
  LocalizableAPIResourceType,
  type APIResource,
} from '../../api'

import deepEqual from '../../utils/deepEqual'
import { Branded } from '../../utils/branded'

import { type Action, ActionTypes } from '../actions'

type CompositeResourceId = Branded<string, 'CompositeResourceId'>

function isValidAPIResourceType(resourceType: string): resourceType is APIResourceType {
  return resourceType in APIResourceType
}

function isLocalizableAPIResourceType(
  resourceType: APIResourceType,
): resourceType is LocalizableAPIResourceType {
  return resourceType in LocalizableAPIResourceType
}

function compositeId(
  resourceId: string,
  resourceType: APIResourceType,
  locale?: string | null,
): CompositeResourceId {
  return (
    locale != null && isLocalizableAPIResourceType(resourceType)
      ? `${resourceId}@${locale}`
      : resourceId
  ) as CompositeResourceId
}

function parseCompositeId(compositeId: CompositeResourceId): [string, string | null] {
  const [resourceId, locale] = compositeId.split('@')
  return [resourceId, locale ?? null]
}

export type State = Map<APIResourceType, Map<CompositeResourceId, APIResource | null>>

export type SerializedState = {
  [key in APIResourceType]?: {
    id: string
    value: Extract<APIResource, { __typename: key }> | null
    locale?: APIResourceLocale<key>
  }[]
}

export function getInitialState(
  serializedState: SerializedState = {
    Swatch: [],
    File: [],
    Typography: [],
    PagePathnameSlice: [],
    GlobalElement: [],
    Table: [],
    Snippet: [],
    Page: [],
    Site: [],
    LocalizedGlobalElement: [],
  },
): State {
  return new Map(
    Object.entries(serializedState).map(([resourceType, resources]) => [
      resourceType,
      new Map(
        resources.map(({ id, value, locale }) => [
          compositeId(id, resourceType as APIResourceType, locale),
          value,
        ]),
      ),
    ]),
  ) as State
}

export function getSerializedState(state: State): SerializedState {
  const resourceMap: SerializedState = {
    Swatch: [],
    File: [],
    Typography: [],
    PagePathnameSlice: [],
    GlobalElement: [],
    Table: [],
    Snippet: [],
    Page: [],
    Site: [],
    LocalizedGlobalElement: [],
  }

  Array.from(state.entries()).forEach(([resourceType, resources]) => {
    const particularResourceMap: { id: string; value: any; locale?: any }[] = []

    Array.from(resources.entries()).forEach(([compositeId, value]) => {
      if (value != null) {
        const [id, locale] = parseCompositeId(compositeId)
        console.assert(
          locale == null || isLocalizableAPIResourceType(resourceType),
          `Unexpected locale for non-localizable resource type ${resourceType}`,
          { id, locale },
        )

        particularResourceMap.push(locale != null ? { id, value, locale } : { id, value })
      }
    })

    resourceMap[resourceType] = particularResourceMap
  })

  return resourceMap
}

export function getHasAPIResource<T extends APIResourceType>(
  state: State,
  resourceType: T,
  resourceId: string,
  locale?: APIResourceLocale<T>,
): boolean {
  return state.get(resourceType)?.has(compositeId(resourceId, resourceType, locale)) ?? false
}

export function getAPIResource<T extends APIResourceType>(
  state: State,
  resourceType: T,
  resourceId: string,
  locale?: APIResourceLocale<T>,
): Extract<APIResource, { __typename: T }> | null {
  const resource = state.get(resourceType)?.get(compositeId(resourceId, resourceType, locale))

  return resource?.__typename === resourceType
    ? (resource as Extract<typeof resource, { __typename: T }>)
    : null
}

export function reducer(state: State = getInitialState(), action: Action): State {
  switch (action.type) {
    case ActionTypes.UPDATE_API_CLIENT_CACHE: {
      const { apiResources } = action.payload

      return Object.entries(apiResources).reduce((state, [resourceType, cachedResources]) => {
        const resType = resourceType as APIResourceType
        const cached:
          | { id: string; value: APIResource | null; locale?: string | null }[]
          | undefined = cachedResources

        const existing = state.get(resType) ?? new Map<CompositeResourceId, APIResource | null>()
        const updated = cached?.reduce((r, { id, value, locale }) => {
          const cid = compositeId(id, resType, locale)
          return r.get(cid) != null ? r : new Map(r).set(cid, value)
        }, existing)

        return updated == null || updated === existing
          ? state
          : new Map(state).set(resType, updated)
      }, state)
    }

    case ActionTypes.API_RESOURCE_FULFILLED: {
      const { resourceType, resourceId, resource, locale } = action.payload
      return new Map(state).set(
        resourceType,
        new Map(state.get(resourceType)!).set(
          compositeId(resourceId, resourceType, locale),
          resource,
        ),
      )
    }

    case ActionTypes.CHANGE_API_RESOURCE: {
      const { resource, locale } = action.payload
      const existingApiResource = getAPIResource(state, resource.__typename, resource.id, locale)

      if (deepEqual(existingApiResource, resource)) return state

      return new Map(state).set(
        resource.__typename,
        new Map(state.get(resource.__typename)!).set(
          compositeId(resource.id, resource.__typename, locale),
          resource,
        ),
      )
    }

    case ActionTypes.EVICT_API_RESOURCE: {
      const { id, locale } = action.payload
      const [resourceType, resourceId] = id.split(':')

      if (!isValidAPIResourceType(resourceType)) return state

      const resources = new Map(state.get(resourceType as APIResourceType)!)

      const deleted = resources.delete(compositeId(resourceId, resourceType, locale))

      return deleted ? new Map(state).set(resourceType as APIResourceType, resources) : state
    }

    default:
      return state
  }
}
