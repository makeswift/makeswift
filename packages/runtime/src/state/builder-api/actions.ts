import { type ThunkAction } from '@reduxjs/toolkit'

import { type SerializedControl } from '../../builder'
import { type PropControllerMessage } from '../../prop-controllers/instances'

import { type Document } from '../modules/read-only-documents'
import { type BoxModel } from '../modules/box-models'
import { type ComponentMeta } from '../modules/components-meta'

import { type Size } from '../react-builder-preview'
import { type DocumentPayload, type SharedAction, SharedActionTypes } from '../shared-api'

export const BuilderActionTypes = {
  ...SharedActionTypes,

  MAKESWIFT_CONNECTION_CHECK: 'MAKESWIFT_CONNECTION_CHECK',

  CHANGE_ELEMENT_BOX_MODELS: 'CHANGE_ELEMENT_BOX_MODELS',

  MOUNT_COMPONENT: 'MOUNT_COMPONENT',
  UNMOUNT_COMPONENT: 'UNMOUNT_COMPONENT',

  CHANGE_DOCUMENT_ELEMENT_SIZE: 'CHANGE_DOCUMENT_ELEMENT_SIZE',
  MESSAGE_BUILDER_PROP_CONTROLLER: 'MESSAGE_BUILDER_PROP_CONTROLLER',

  HANDLE_WHEEL: 'HANDLE_WHEEL',
  HANDLE_POINTER_MOVE: 'HANDLE_POINTER_MOVE',

  ELEMENT_FROM_POINT_CHANGE: 'ELEMENT_FROM_POINT_CHANGE',

  REGISTER_BUILDER_DOCUMENT: 'REGISTER_BUILDER_DOCUMENT',
  UNREGISTER_BUILDER_DOCUMENT: 'UNREGISTER_BUILDER_DOCUMENT',

  REGISTER_BUILDER_COMPONENT: 'REGISTER_BUILDER_COMPONENT',
  UNREGISTER_BUILDER_COMPONENT: 'UNREGISTER_BUILDER_COMPONENT',

  HANDLE_HOST_NAVIGATE: 'HANDLE_HOST_NAVIGATE',
} as const

type ActionWithTransferables<A> = A & {
  transferables?: Transferable[]
}

type MakeswiftConnectionCheckAction = {
  type: typeof BuilderActionTypes.MAKESWIFT_CONNECTION_CHECK
  payload: { currentUrl: string }
}

type ChangeElementBoxModelsAction = {
  type: typeof BuilderActionTypes.CHANGE_ELEMENT_BOX_MODELS
  payload: { changedElementBoxModels: Map<string, Map<string, BoxModel | null>> }
}

type MountComponentAction = {
  type: typeof BuilderActionTypes.MOUNT_COMPONENT
  payload: { documentKey: string; elementKey: string }
}

type UnmountComponentAction = {
  type: typeof BuilderActionTypes.UNMOUNT_COMPONENT
  payload: { documentKey: string; elementKey: string }
}

type ChangeDocumentElementSizeAction = {
  type: typeof BuilderActionTypes.CHANGE_DOCUMENT_ELEMENT_SIZE
  payload: { size: Size }
}

type MessageBuilderPropControllerAction<T = PropControllerMessage> = {
  type: typeof BuilderActionTypes.MESSAGE_BUILDER_PROP_CONTROLLER
  payload: { documentKey: string; elementKey: string; propName: string; message: T }
}

type HandleWheelAction = {
  type: typeof BuilderActionTypes.HANDLE_WHEEL
  payload: { deltaX: number; deltaY: number }
}

type HandlePointerMoveAction = {
  type: typeof BuilderActionTypes.HANDLE_POINTER_MOVE
  payload: { clientX: number; clientY: number }
}

type ElementFromPointChangeAction = {
  type: typeof BuilderActionTypes.ELEMENT_FROM_POINT_CHANGE
  payload: { keys: { documentKey: string; elementKey: string } | null }
}

type RegisterBuilderDocumentAction = {
  type: typeof BuilderActionTypes.REGISTER_BUILDER_DOCUMENT
  payload: { documentKey: string; document: DocumentPayload }
}

type UnregisterBuilderDocumentAction = {
  type: typeof BuilderActionTypes.UNREGISTER_BUILDER_DOCUMENT
  payload: { documentKey: string }
}

type RegisterBuilderComponentAction = {
  type: typeof BuilderActionTypes.REGISTER_BUILDER_COMPONENT
  payload: {
    type: string
    meta: ComponentMeta
    serializedControls: Record<string, SerializedControl>
  }
}

type UnregisterBuilderComponentAction = {
  type: typeof BuilderActionTypes.UNREGISTER_BUILDER_COMPONENT
  payload: { type: string }
}

type HandleHostNavigateAction = {
  type: typeof BuilderActionTypes.HANDLE_HOST_NAVIGATE
  payload: { url: string | null }
}

export type BuilderAction =
  | SharedAction
  | MakeswiftConnectionCheckAction
  | ChangeElementBoxModelsAction
  | MountComponentAction
  | UnmountComponentAction
  | ChangeDocumentElementSizeAction
  | MessageBuilderPropControllerAction
  | HandleWheelAction
  | HandlePointerMoveAction
  | ElementFromPointChangeAction
  | RegisterBuilderDocumentAction
  | UnregisterBuilderDocumentAction
  | RegisterBuilderComponentAction
  | UnregisterBuilderComponentAction
  | HandleHostNavigateAction

export function makeswiftConnectionCheck(
  payload: MakeswiftConnectionCheckAction['payload'],
): MakeswiftConnectionCheckAction {
  return { type: BuilderActionTypes.MAKESWIFT_CONNECTION_CHECK, payload }
}

export function changeElementBoxModels(
  changedElementBoxModels: Map<string, Map<string, BoxModel | null>>,
): ChangeElementBoxModelsAction {
  return {
    type: BuilderActionTypes.CHANGE_ELEMENT_BOX_MODELS,
    payload: { changedElementBoxModels },
  }
}

export function mountComponent(documentKey: string, elementKey: string): MountComponentAction {
  return { type: BuilderActionTypes.MOUNT_COMPONENT, payload: { documentKey, elementKey } }
}

export function unmountComponent(documentKey: string, elementKey: string): UnmountComponentAction {
  return { type: BuilderActionTypes.UNMOUNT_COMPONENT, payload: { documentKey, elementKey } }
}

export function mountComponentEffect(
  documentKey: string,
  elementKey: string,
): ThunkAction<() => void, unknown, unknown, BuilderAction> {
  return dispatch => {
    dispatch(mountComponent(documentKey, elementKey))

    return () => {
      dispatch(unmountComponent(documentKey, elementKey))
    }
  }
}

export function changeDocumentElementSize(size: Size): ChangeDocumentElementSizeAction {
  return { type: BuilderActionTypes.CHANGE_DOCUMENT_ELEMENT_SIZE, payload: { size } }
}

export function messageBuilderPropController<T>(
  documentKey: string,
  elementKey: string,
  propName: string,
  message: T,
): MessageBuilderPropControllerAction<T> {
  return {
    type: BuilderActionTypes.MESSAGE_BUILDER_PROP_CONTROLLER,
    payload: { documentKey, elementKey, propName, message },
  }
}

export function handleWheel(payload: { deltaX: number; deltaY: number }): HandleWheelAction {
  return { type: BuilderActionTypes.HANDLE_WHEEL, payload }
}

export function handlePointerMove(payload: {
  clientX: number
  clientY: number
}): HandlePointerMoveAction {
  return { type: BuilderActionTypes.HANDLE_POINTER_MOVE, payload }
}

export function elementFromPointChange(
  keys: {
    documentKey: string
    elementKey: string
  } | null,
): ElementFromPointChangeAction {
  return { type: BuilderActionTypes.ELEMENT_FROM_POINT_CHANGE, payload: { keys } }
}

export function registerBuilderDocument(document: Document): RegisterBuilderDocumentAction {
  return {
    type: BuilderActionTypes.REGISTER_BUILDER_DOCUMENT,
    payload: { documentKey: document.key, document },
  }
}

export function unregisterBuilderDocument(documentKey: string): UnregisterBuilderDocumentAction {
  return { type: BuilderActionTypes.UNREGISTER_BUILDER_DOCUMENT, payload: { documentKey } }
}

export function registerBuilderDocumentsEffect(
  documents: Document[],
): ThunkAction<() => void, unknown, unknown, BuilderAction> {
  return dispatch => {
    documents.forEach(document => dispatch(registerBuilderDocument(document)))

    return () => {
      documents.forEach(document => dispatch(unregisterBuilderDocument(document.key)))
    }
  }
}

export function registerBuilderComponent(
  payload: RegisterBuilderComponentAction['payload'],
  transferables?: Transferable[],
): ActionWithTransferables<RegisterBuilderComponentAction> {
  return {
    type: BuilderActionTypes.REGISTER_BUILDER_COMPONENT,
    payload,
    transferables,
  }
}

export function unregisterBuilderComponent(
  payload: UnregisterBuilderComponentAction['payload'],
): UnregisterBuilderComponentAction {
  return { type: BuilderActionTypes.UNREGISTER_BUILDER_COMPONENT, payload }
}

export function hasTransferables<A extends BuilderAction>(
  action: A,
): action is ActionWithTransferables<A> {
  return 'transferables' in action && Array.isArray(action.transferables)
}
