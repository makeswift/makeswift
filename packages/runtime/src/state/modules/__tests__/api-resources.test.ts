import * as APIResources from '../api-resources'
import * as Actions from '../../actions'
import {
  APIResourceType,
  type Swatch,
  type File,
  type Typography,
  type PagePathnameSlice,
  type GlobalElement,
  type LocalizedGlobalElement,
  type Table,
  type Snippet,
  type Page,
  type Site,
  type APIResource,
} from '../../../api'

const swatch: Swatch = {
  __typename: APIResourceType.Swatch,
  id: '[test-swatch-id]',
  hue: 100,
  saturation: 70,
  lightness: 90,
}

const file: File = {
  __typename: APIResourceType.File,
  id: '[test-file-id]',
  name: 'test-file-name',
  extension: 'png',
  publicUrl: 'https://example.com/test-file.png',
  dimensions: { width: 1024, height: 1024 },
}

const typography: Typography = {
  __typename: APIResourceType.Typography,
  id: '[test-typography-id]',
  name: 'test-typography-name',
  style: [],
}

const pagePathnameSlice: PagePathnameSlice = {
  __typename: APIResourceType.PagePathnameSlice,
  id: '[test-page-pathname-slice-id]',
  pathname: 'test/pathname',
}

const localizedPagePathnameSlice_fr_FR: PagePathnameSlice = {
  __typename: APIResourceType.PagePathnameSlice,
  id: '[test-page-pathname-slice-id]',
  pathname: 'test-pathname',
  basePageId: '[test-base-page-id]',
  localizedPathname: 'fr-FR/test/localized-pathname',
}

const localizedPagePathnameSlice_it_IT: PagePathnameSlice = {
  __typename: APIResourceType.PagePathnameSlice,
  id: '[test-page-pathname-slice-id]',
  pathname: 'test-pathname',
  basePageId: '[test-base-page-id]',
  localizedPathname: 'it-IT/test/localized-pathname',
}

const globalElement: GlobalElement = {
  __typename: APIResourceType.GlobalElement,
  id: '[test-global-element-id]',
  data: {},
}

const localizedGlobalElement_fr_FR: LocalizedGlobalElement = {
  __typename: APIResourceType.LocalizedGlobalElement,
  id: '[test-localized-global-element-id]',
  data: { prompt: 'bonjour le monde' },
}

const localizedGlobalElement_it_IT: LocalizedGlobalElement = {
  __typename: APIResourceType.LocalizedGlobalElement,
  id: '[test-localized-global-element-id]',
  data: { prompt: 'ciao mondo' },
}

const table: Table = {
  __typename: APIResourceType.Table,
  id: '[test-table-id]',
  name: 'test-table-name',
  columns: [],
}

const snippet: Snippet = {
  __typename: APIResourceType.Snippet,
  id: '[test-snippet-id]',
  name: 'test-snippet-name',
  code: 'test-snippet-code',
  cleanup: null,
  location: 'BODY',
  shouldAddToNewPages: false,
  liveEnabled: false,
  builderEnabled: false,
}

const page: Page = {
  __typename: APIResourceType.Page,
  id: '[test-page-id]',
  snippets: [],
}

const site: Site = {
  __typename: APIResourceType.Site,
  id: '[test-site-id]',
  googleFonts: {
    edges: [],
  },
}

describe('state / APIResources', () => {
  test('`getInitialState` w/o arguments returns an empty state', () => {
    const state = APIResources.getInitialState()
    const serializedState = APIResources.getSerializedState(state)
    expect(serializedState).toMatchSnapshot()
  })

  test('`getInitialState`/`getSerializedState` are symmetrical', () => {
    const serializedState = {
      Swatch: [{ id: swatch.id, value: swatch }],
      File: [{ id: file.id, value: file }],
      Typography: [{ id: typography.id, value: typography }],
      PagePathnameSlice: [
        { id: pagePathnameSlice.id, value: pagePathnameSlice },
        {
          id: localizedPagePathnameSlice_fr_FR.id,
          value: localizedPagePathnameSlice_fr_FR,
          locale: 'fr-FR',
        },
        {
          id: localizedPagePathnameSlice_it_IT.id,
          value: localizedPagePathnameSlice_it_IT,
          locale: 'it-IT',
        },
      ],
      GlobalElement: [{ id: globalElement.id, value: globalElement }],
      LocalizedGlobalElement: [
        {
          id: localizedGlobalElement_fr_FR.id,
          value: localizedGlobalElement_fr_FR,
          locale: 'fr-FR',
        },
        {
          id: localizedGlobalElement_it_IT.id,
          value: localizedGlobalElement_it_IT,
          locale: 'it-IT',
        },
      ],
      Table: [{ id: table.id, value: table }],
      Snippet: [{ id: snippet.id, value: snippet }],
      Page: [{ id: page.id, value: page }],
      Site: [{ id: site.id, value: site }],
    }

    const state = APIResources.getInitialState(serializedState)
    expect(APIResources.getSerializedState(state)).toStrictEqual(serializedState)
  })

  test('`Actions.apiResourceFulfilled` correctly populates the store', () => {
    const state = (
      [
        [swatch, null],
        [file, null],
        [typography, null],
        [pagePathnameSlice, null],
        [localizedPagePathnameSlice_fr_FR, 'fr-FR'],
        [globalElement, null],
        [localizedGlobalElement_fr_FR, 'fr-FR'],
        [localizedGlobalElement_it_IT, 'it-IT'],
        [table, null],
        [snippet, null],
        [page, null],
        [site, null],
      ] as [APIResource, string][]
    ).reduce(
      (state, [resource, locale]) =>
        APIResources.reducer(
          state,
          Actions.apiResourceFulfilled(resource.__typename, resource.id, resource, locale),
        ),
      APIResources.getInitialState(),
    )

    expect(APIResources.getSerializedState(state)).toMatchSnapshot()
  })

  test('`Actions.changeApiResource` correctly updates the store', () => {
    const state = (
      [
        [swatch, null],
        [pagePathnameSlice, null],
        [localizedPagePathnameSlice_fr_FR, 'fr-FR'],
        [localizedPagePathnameSlice_it_IT, 'it-IT'],
        [globalElement, null],
        [localizedGlobalElement_fr_FR, 'fr-FR'],
        [localizedGlobalElement_it_IT, 'it-IT'],
      ] as [APIResource, string][]
    ).reduce(
      (state, [resource, locale]) =>
        APIResources.reducer(
          state,
          Actions.apiResourceFulfilled(resource.__typename, resource.id, resource, locale),
        ),
      APIResources.getInitialState(),
    )

    const updatedState = (
      [
        [{ ...swatch, hue: 17 }, null],
        [{ ...pagePathnameSlice, pathname: 'test/new-pathname' }, null],
        [
          {
            ...localizedPagePathnameSlice_it_IT,
            pathname: 'test/new-pathname',
            localizedPathname: 'test/new-localized-pathname-it-IT',
          },
          'it-IT',
        ],
        [{ ...localizedGlobalElement_fr_FR, data: { prompt: 'hello world' } }, 'fr-FR'],
      ] as [APIResource, string][]
    ).reduce(
      (state, [resource, locale]) =>
        APIResources.reducer(state, Actions.changeApiResource(resource, locale)),
      state,
    )

    expect(APIResources.getSerializedState(updatedState)).toMatchSnapshot()
  })

  test('`Actions.evictApiResource` correctly removes the resource from the store', () => {
    const state = (
      [
        [swatch, null],
        [pagePathnameSlice, null],
        [localizedPagePathnameSlice_fr_FR, 'fr-FR'],
        [localizedPagePathnameSlice_it_IT, 'it-IT'],
        [globalElement, null],
        [localizedGlobalElement_fr_FR, 'fr-FR'],
        [localizedGlobalElement_it_IT, 'it-IT'],
      ] as [APIResource, string][]
    ).reduce(
      (state, [resource, locale]) =>
        APIResources.reducer(
          state,
          Actions.apiResourceFulfilled(resource.__typename, resource.id, resource, locale),
        ),
      APIResources.getInitialState(),
    )

    const updatedState = (
      [
        [`${APIResourceType.Swatch}:${swatch.id}`, null],
        [`${APIResourceType.PagePathnameSlice}:${localizedPagePathnameSlice_fr_FR.id}`, 'fr-FR'],
        [`${APIResourceType.LocalizedGlobalElement}:${localizedGlobalElement_it_IT.id}`, 'it-IT'],
      ] as [string, string][]
    ).reduce(
      (state, [id, locale]) => APIResources.reducer(state, Actions.evictApiResource(id, locale)),
      state,
    )

    expect(APIResources.getSerializedState(updatedState)).toMatchSnapshot()
  })

  describe('Only getters for localizable resources accept the locale argument', () => {
    const state = APIResources.getInitialState()
    describe.each([
      APIResourceType.Swatch,
      APIResourceType.File,
      APIResourceType.Typography,
      APIResourceType.GlobalElement,
      APIResourceType.Table,
      APIResourceType.Snippet,
      APIResourceType.Page,
      APIResourceType.Site,
    ])('getter for %s', resourceType => {
      test.each(['fr-FR', null, undefined])('fails to compile with locale %s', locale => {
        //@ts-expect-error
        expect(APIResources.getHasAPIResource(state, resourceType, '', locale)).toBe(false)
        //@ts-expect-error
        expect(APIResources.getAPIResource(state, resourceType, '', locale)).toBe(null)
      })

      test('compiles w/o the locale argument', () => {
        expect(APIResources.getHasAPIResource(state, resourceType, '')).toBe(false)
        expect(APIResources.getAPIResource(state, resourceType, '')).toBe(null)
      })
    })

    describe.each([APIResourceType.LocalizedGlobalElement, APIResourceType.PagePathnameSlice])(
      'getter for %s',
      resourceType => {
        test.each(['fr-FR', null, undefined])('compiles w/ locale %s', locale => {
          expect(APIResources.getHasAPIResource(state, resourceType, '', locale)).toBe(false)
          expect(APIResources.getAPIResource(state, resourceType, '', locale)).toBe(null)
        })
      },
    )
  })

  test('`Actions.updateAPIClientCache` fills in the state from cache w/o overriding existing resources', () => {
    const state = (
      [
        [swatch, null],
        [pagePathnameSlice, null],
        [localizedPagePathnameSlice_fr_FR, 'fr-FR'],
        [globalElement, null],
        [localizedGlobalElement_it_IT, 'it-IT'],
      ] as [APIResource, string][]
    ).reduce(
      (state, [resource, locale]) =>
        APIResources.reducer(
          state,
          Actions.apiResourceFulfilled(resource.__typename, resource.id, resource, locale),
        ),
      APIResources.getInitialState(),
    )

    const apiResources = {
      Swatch: [{ id: swatch.id, value: { ...swatch, hue: 17 } }],
      PagePathnameSlice: [
        {
          id: pagePathnameSlice.id,
          value: { ...pagePathnameSlice, pathname: 'test/new-pathname' },
        },
        {
          id: localizedPagePathnameSlice_fr_FR.id,
          value: {
            ...localizedPagePathnameSlice_fr_FR,
            localizedPathname: 'test/new-localized-pathname-fr-FR',
          },
          locale: 'fr-FR',
        },
        {
          id: localizedPagePathnameSlice_it_IT.id,
          value: localizedPagePathnameSlice_it_IT,
          locale: 'it-IT',
        },
      ],
      LocalizedGlobalElement: [
        {
          id: localizedGlobalElement_fr_FR.id,
          value: localizedGlobalElement_fr_FR,
          locale: 'fr-FR',
        },
        {
          id: localizedGlobalElement_it_IT.id,
          value: {
            ...localizedGlobalElement_it_IT,
            localizedPathname: 'test/new-localized-pathname-it-IT',
          },
          locale: 'it-IT',
        },
      ],
      Table: [{ id: table.id, value: table }],
      Snippet: [{ id: snippet.id, value: snippet }],
    }

    const updatedState = APIResources.reducer(
      state,
      Actions.updateAPIClientCache({ apiResources, localizedResourcesMap: {} }),
    )

    expect(APIResources.getSerializedState(updatedState)).toMatchSnapshot()
  })
})
