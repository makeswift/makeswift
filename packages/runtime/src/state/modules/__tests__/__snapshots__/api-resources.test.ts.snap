// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state / APIResources \`Actions.apiResourceFulfilled\` correctly populates the store 1`] = `
{
  "File": [
    {
      "id": "[test-file-id]",
      "value": {
        "__typename": "File",
        "dimensions": {
          "height": 1024,
          "width": 1024,
        },
        "extension": "png",
        "id": "[test-file-id]",
        "name": "test-file-name",
        "publicUrl": "https://example.com/test-file.png",
      },
    },
  ],
  "GlobalElement": [
    {
      "id": "[test-global-element-id]",
      "value": {
        "__typename": "GlobalElement",
        "data": {},
        "id": "[test-global-element-id]",
      },
    },
  ],
  "LocalizedGlobalElement": [
    {
      "id": "[test-localized-global-element-id]",
      "locale": "fr-FR",
      "value": {
        "__typename": "LocalizedGlobalElement",
        "data": {
          "prompt": "bonjour le monde",
        },
        "id": "[test-localized-global-element-id]",
      },
    },
    {
      "id": "[test-localized-global-element-id]",
      "locale": "it-IT",
      "value": {
        "__typename": "LocalizedGlobalElement",
        "data": {
          "prompt": "ciao mondo",
        },
        "id": "[test-localized-global-element-id]",
      },
    },
  ],
  "Page": [
    {
      "id": "[test-page-id]",
      "value": {
        "__typename": "Page",
        "id": "[test-page-id]",
        "snippets": [],
      },
    },
  ],
  "PagePathnameSlice": [
    {
      "id": "[test-page-pathname-slice-id]",
      "value": {
        "__typename": "PagePathnameSlice",
        "id": "[test-page-pathname-slice-id]",
        "pathname": "test/pathname",
      },
    },
    {
      "id": "[test-page-pathname-slice-id]",
      "locale": "fr-FR",
      "value": {
        "__typename": "PagePathnameSlice",
        "basePageId": "[test-base-page-id]",
        "id": "[test-page-pathname-slice-id]",
        "localizedPathname": "fr-FR/test/localized-pathname",
        "pathname": "test-pathname",
      },
    },
  ],
  "Site": [
    {
      "id": "[test-site-id]",
      "value": {
        "__typename": "Site",
        "googleFonts": {
          "edges": [],
        },
        "id": "[test-site-id]",
      },
    },
  ],
  "Snippet": [
    {
      "id": "[test-snippet-id]",
      "value": {
        "__typename": "Snippet",
        "builderEnabled": false,
        "cleanup": null,
        "code": "test-snippet-code",
        "id": "[test-snippet-id]",
        "liveEnabled": false,
        "location": "BODY",
        "name": "test-snippet-name",
        "shouldAddToNewPages": false,
      },
    },
  ],
  "Swatch": [
    {
      "id": "[test-swatch-id]",
      "value": {
        "__typename": "Swatch",
        "hue": 100,
        "id": "[test-swatch-id]",
        "lightness": 90,
        "saturation": 70,
      },
    },
  ],
  "Table": [
    {
      "id": "[test-table-id]",
      "value": {
        "__typename": "Table",
        "columns": [],
        "id": "[test-table-id]",
        "name": "test-table-name",
      },
    },
  ],
  "Typography": [
    {
      "id": "[test-typography-id]",
      "value": {
        "__typename": "Typography",
        "id": "[test-typography-id]",
        "name": "test-typography-name",
        "style": [],
      },
    },
  ],
}
`;

exports[`state / APIResources \`Actions.changeApiResource\` correctly updates the store 1`] = `
{
  "File": [],
  "GlobalElement": [
    {
      "id": "[test-global-element-id]",
      "value": {
        "__typename": "GlobalElement",
        "data": {},
        "id": "[test-global-element-id]",
      },
    },
  ],
  "LocalizedGlobalElement": [
    {
      "id": "[test-localized-global-element-id]",
      "locale": "fr-FR",
      "value": {
        "__typename": "LocalizedGlobalElement",
        "data": {
          "prompt": "hello world",
        },
        "id": "[test-localized-global-element-id]",
      },
    },
    {
      "id": "[test-localized-global-element-id]",
      "locale": "it-IT",
      "value": {
        "__typename": "LocalizedGlobalElement",
        "data": {
          "prompt": "ciao mondo",
        },
        "id": "[test-localized-global-element-id]",
      },
    },
  ],
  "Page": [],
  "PagePathnameSlice": [
    {
      "id": "[test-page-pathname-slice-id]",
      "value": {
        "__typename": "PagePathnameSlice",
        "id": "[test-page-pathname-slice-id]",
        "pathname": "test/new-pathname",
      },
    },
    {
      "id": "[test-page-pathname-slice-id]",
      "locale": "fr-FR",
      "value": {
        "__typename": "PagePathnameSlice",
        "basePageId": "[test-base-page-id]",
        "id": "[test-page-pathname-slice-id]",
        "localizedPathname": "fr-FR/test/localized-pathname",
        "pathname": "test-pathname",
      },
    },
    {
      "id": "[test-page-pathname-slice-id]",
      "locale": "it-IT",
      "value": {
        "__typename": "PagePathnameSlice",
        "basePageId": "[test-base-page-id]",
        "id": "[test-page-pathname-slice-id]",
        "localizedPathname": "test/new-localized-pathname-it-IT",
        "pathname": "test/new-pathname",
      },
    },
  ],
  "Site": [],
  "Snippet": [],
  "Swatch": [
    {
      "id": "[test-swatch-id]",
      "value": {
        "__typename": "Swatch",
        "hue": 17,
        "id": "[test-swatch-id]",
        "lightness": 90,
        "saturation": 70,
      },
    },
  ],
  "Table": [],
  "Typography": [],
}
`;

exports[`state / APIResources \`Actions.evictApiResource\` correctly removes the resource from the store 1`] = `
{
  "File": [],
  "GlobalElement": [
    {
      "id": "[test-global-element-id]",
      "value": {
        "__typename": "GlobalElement",
        "data": {},
        "id": "[test-global-element-id]",
      },
    },
  ],
  "LocalizedGlobalElement": [
    {
      "id": "[test-localized-global-element-id]",
      "locale": "fr-FR",
      "value": {
        "__typename": "LocalizedGlobalElement",
        "data": {
          "prompt": "bonjour le monde",
        },
        "id": "[test-localized-global-element-id]",
      },
    },
  ],
  "Page": [],
  "PagePathnameSlice": [
    {
      "id": "[test-page-pathname-slice-id]",
      "value": {
        "__typename": "PagePathnameSlice",
        "id": "[test-page-pathname-slice-id]",
        "pathname": "test/pathname",
      },
    },
    {
      "id": "[test-page-pathname-slice-id]",
      "locale": "it-IT",
      "value": {
        "__typename": "PagePathnameSlice",
        "basePageId": "[test-base-page-id]",
        "id": "[test-page-pathname-slice-id]",
        "localizedPathname": "it-IT/test/localized-pathname",
        "pathname": "test-pathname",
      },
    },
  ],
  "Site": [],
  "Snippet": [],
  "Swatch": [],
  "Table": [],
  "Typography": [],
}
`;

exports[`state / APIResources \`getInitialState\` w/o arguments returns an empty state 1`] = `
{
  "File": [],
  "GlobalElement": [],
  "LocalizedGlobalElement": [],
  "Page": [],
  "PagePathnameSlice": [],
  "Site": [],
  "Snippet": [],
  "Swatch": [],
  "Table": [],
  "Typography": [],
}
`;
